File: simplified-books-api/src/main/java/org/nypl/simplified/books/api/BookLocation.java
package org.nypl.simplified.books.api;

import com.google.auto.value.AutoValue;
import com.io7m.jfunctional.OptionType;

import java.io.Serializable;

/**
 * The current page. A specific location in an EPUB is identified by an
 * <i>idref</i> and a <i>content CFI</i>. In some cases, the <i>content CFI</i>
 * may not be present.
 *
 * <p>Note: The type is {@link Serializable} purely because the Android API requires this
 * in order pass values of this type between activities. We make absolutely no guarantees
 * that serialized values of this class will be compatible with future releases.</p>
 */

@AutoValue
public abstract class BookLocation implements Serializable {

  /**
   * @return The content CFI, if any
   */

  public abstract OptionType<String> contentCFI();

  /**
   * @return The IDRef
   */

  public abstract String idRef();

  /**
   * Create a book location.
   *
   * @param cfi    The content CFI, if any
   * @param id_ref The IDRef
   * @return A book location
   */

  public static BookLocation create(
      final OptionType<String> cfi,
      final String id_ref) {
    return new AutoValue_BookLocation(cfi, id_ref);
  }
}

File: simplified-books-api/src/main/java/org/nypl/simplified/books/api/BookLocation.kt
package org.nypl.simplified.books.api

import java.io.Serializable

/**
 * The current page. A specific location in an EPUB is identified by a chapter index, or an
 * *idref* and a *content CFI*. In some cases, the *content CFI*
 * may not be present.
 *
 * Note: The type is [Serializable] purely because the Android API requires this
 * in order pass values of this type between activities. We make absolutely no guarantees
 * that serialized values of this class will be compatible with future releases.
 */

sealed class BookLocation : Serializable {

  /**
   * R2-specific book locations.
   */

  data class BookLocationR2(
    val progress: BookChapterProgress
  ) : BookLocation()

  /**
   * R1-specific book locations.
   */

  @Deprecated("Use R2")
  data class BookLocationR1(

    /**
     * The chapter progress.
     */

    val progress: Double?,

    /**
     * @return The content CFI, if any
     */

    val contentCFI: String?,

    /**
     * @return The IDRef
     */

    val idRef: String?
  ) : BookLocation()
}
